### PR TITLE
Add Deepfake Detection and Dockerization Support; Improved Image Splicing and Copy Move Detection

### PR DESCRIPTION
**Pull request by d4nzk**
**EDIT: This modified version of Sherloq was tested in Python 3.12**

## Overview
This PR introduces significant improvements to Sherloq, adding a new Deepfake Detection module and Docker support for easier deployment and usage, also improving the existing Image Splicing and Copy Move Detection modules.

## What's New?
1. **Image Splicing Improvements**
- How it works: An unsupervised pipeline that doesn't rely on pre-trained CNNs. It breaks the image into SLIC Superpixels and analyzes them through 5 independent mathematical "lenses" (High-Frequency Noise, DCT, FFT, ELA, and Color Variance), then it uses PCA to automatically weight the 5 lenses and K-Means (k=2) to automatically find the perfect threshold.
- UI Updates: Added dynamic sliders for real-time tweaking, backed by a "Smart Shape Filter" to automatically ignore false positives.

2. **Copy-Move Detection Improvements**
- How it works: A robust keypoint-matching approach inspired by SMDAF. It extracts keypoints from both the original and a 2x scaled version of the image for better accuracy. It uses Lowe's Ratio and Double Adaptive Filtering to verify matches, RANSAC to confirm geometric relationships, and Multi-DBSCAN to cluster the cloned regions.
- UI Updates: Changed the control interface by replacing the static dropdown menu with a dynamic slider for smoother, more intuitive parameter adjustments. Draws visual convex hulls and connecting lines over the cloned areas.

3. **Deepfake Detection Module**
- How it works: A Vision Transformer (ViT) classification module (Deep-Fake-Detector-v2-Model from Hugging Face). It Splits a 224x224 image into 16x16 patches, using self-attention to spot visual inconsistencies, blending errors, and AI artifacts to classify the image as Real or Deepfake.
- UI Updates: Added a real-time progress bar to track the Vision Transformer's inference status, providing clear visual feedback during heavy processing before displaying the final Real/Fake classification results.

4. **Docker Support**
- The Goal: Make Sherloq easy to deploy on Windows via headless X11 forwarding (VcXsrv) without dependency nightmares.
- The Optimization: Shaved the Docker image down to ~7.8 GB by explicitly forcing pip to install CPU-only versions of heavy AI libraries (like PyTorch), preventing the accidental 4 GB NVIDIA CUDA download. Chained apt-get and pip commands to instantly clear installation caches.

5. **Updated README.md**
- Added how to run Dockerfile in the "Running with Docker" section and how to add a new module in the "Adding a New Module" section

## Testing Performed
- [x] Copy-Move Performance Validation: Successfully mitigated high memory usage spikes by implementing an automatic image resizer (scaling down images >1920px) and refining keypoint filtering and Multi-DBSCAN parameters.
- [x] Deepfake Model Benchmarking: The ViT model achieved a 66.72% Accuracy and 0.686 F1-Score during evaluation. In live testing, it successfully classified 74 out of 100 mixed images, correctly identifying real datasets (e.g., FFHQ-Real) versus generated ones.
- [x] Image Splicing Evaluation: Benchmarked the pipeline against 100 images (50 authentic, 50 tampered) from the Columbia Uncompressed and IMD2020 datasets. Visually validated the generated consensus heatmaps against the dataset's binary ground truths (edgemask_3).
- [x] Docker & GUI Forwarding: Successfully built and deployed the container utilizing docker run -it --rm -e DISPLAY=host.docker.internal:0.0. Verified that the Sherloq UI successfully forwards to the Windows host via VcXsrv with "Disable Access Control" activated, and verified volume mapping for local dataset testing.

## UI Showcase
- **Image Splicing Detection**
image

- **Copy-Move Detection**
image

- **Deepfake Detection**
image

Waiting for the process

image

After The Process are done

### CODE REVIEW

**Review by GuidoBartoli — COMMENTED** [review]
Why did you remove the putText() line for showing "Splicing detected"?

**Comment by GuidoBartoli** [comment]
Nice PR, really thanks for all your effort!
Did you also test it under Linux environment?

**Comment by d4nzk** [comment]
I personally didn't test it on linux, but my friends who helped me in making this version of Sherloq did. They said that it worked fine, but i didn't personally test it myself
Edit: i also need to add that in the requirements.txt, i added these lines
```
python-magic-bin==0.4.* ; sys_platform == 'win32' or sys_platform == 'darwin'
python-magic==0.4.* ; sys_platform == 'linux'
```
so `python-magic` is the one installed while on linux, and `python-magic-bin` is the one installed while on windows and mac (even though i can't really say does it work in mac or not because nobody in my team have a mac device while working on the project)

**Comment by d4nzk** [comment]
> Why did you remove the putText() line for showing "Splicing detected"?

when i tried using a smaller dimension picture, the text will be very big, so that's why i decide to just remove it.
